### DISCUSSION TITLE
libpng causes GUI apps to crash with EXC_BAD_ACCESS

### DISCUSSION THREAD

**Discussion by PythonCoderAS** [Homebrew/discussions#5420] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.3.2-56-gfe16e9d
ORIGIN: https://github.com/Homebrew/brew
HEAD: fe16e9d7ef393fdcf353a3c3eba73d0299b744cc
Last commit: 2 hours ago
Core tap HEAD: 4a00cafc112d78970de81aa080e82c9c6c3ac9f6
Core tap last commit: 9 days ago
Core tap JSON: 31 May 20:41 UTC
Core cask tap HEAD: 01ea5381848b5f69e2e495b86f329433f2caad8f
Core cask tap last commit: 9 days ago
Core cask tap branch: add-blitzgg
Core cask tap JSON: 31 May 20:41 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: /opt/homebrew/bin/nvim
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.2 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.2/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.4-arm64
CLT: 15.3.0.0.1.1708646388
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1
  python@3.8
```

### Description of issue

`libpng` causes some GUI applications that call into Objective C to crash, such as https://github.com/tauri-apps/tauri/issues/7351. The problem goes away whenever I uninstall `libpng` or remove `/opt/homebrew/lib` from `DYLD_LIBRARY_PATH`.

**Comment by carlocab** [comment]
Why are you setting `DYLD_LIBRARY_PATH`?

**Reply by PythonCoderAS** [reply]
Something would not compile without it set.

**Reply by jcelerier** [reply]
Also note that IDEs will set DYLD_LIBRARY_PATH automatically to the path where libraries are found

**Reply by carlocab** [reply]
Using `DYLD_LIBRARY_PATH` that way is a hack, and breaking stuff when using it should be expected. It's also a vector for [dylib highjacking](https://attack.mitre.org/techniques/T1574/004/).

This is why Apple strips `DYLD_LIBRARY_PATH` when running the binaries they provide:
```
# We can pass `DYLD_LIBRARY_PATH` to binaries not built by Apple
❯ DYLD_LIBRARY_PATH=/opt/homebrew/lib /opt/homebrew/bin/bash -c '[[ -n $DYLD_LIBRARY_PATH ]]'; echo $?
0

# We _cannot_ pass `DYLD_LIBRARY_PATH` to binaries built by Apple
❯ DYLD_LIBRARY_PATH=/opt/homebrew/lib /bin/bash -c '[[ -n $DYLD_LIBRARY_PATH ]]'; echo $?
1

# Other environment variables can be passed as normal for both
❯ NOT_DYLD_LIBRARY_PATH=/opt/homebrew/lib /opt/homebrew/bin/bash -c '[[ -n $NOT_DYLD_LIBRARY_PATH ]]'; echo $?
0
❯ NOT_DYLD_LIBRARY_PATH=/opt/homebrew/lib /bin/bash -c '[[ -n $NOT_DYLD_LIBRARY_PATH ]]'; echo $?
0
```

I would strongly advise against the use of any application that sets `DYLD_LIBRARY_PATH` outside of a carefully controlled environment. You probably should not set it yourself either unless you know what you're doing and are very careful.

**Reply by jcelerier** [reply]
so I asked this question for the IDE I use - https://bugreports.qt.io/browse/QTCREATORBUG-31656 

Is there any other way than what's proposed?  

> An app needs to find its runtime libraries somehow. The three possible ways I am aware of are:
> 1. Use rpath
> 2. Deploy the libs/frameworks into the bundle
> 3. Use the DYLD_* variables
> I did not see anything in the linked threads that indicates that people do the first or the second. What do your projects do?

What's the solution ? for sure when developping you want to be able to build and run an app without having to implement all the dylib and install_name_tool logic for every library you're going to use, what would be the best solution in that case ?

**Reply by jcelerier** [reply]
Up @carlocab

**Comment by mikkeljohnsen** [comment]
Is it not possible to install "libpng.dylib" without app crashes (on macOS Seqouia) ?

I need "libpng" for Gtk (gdk-pixbuf, cairo etc). But the app crashes because "libpng.dylib" is installed. If I uninstall "libpng.dylib" it works. But can't load PNG images, so not a solution.

(I'm using dotnet and that must invoke something that fails when libpng is installed.

**Reply by carlocab** [reply]
It should be if your GUI apps are built correctly. They're crashing because they're loading random libraries it finds on your system with no guarantee that they're compatible.

**Reply by mikkeljohnsen** [reply]
All my libraries is compiled with full path to other libs. Sp no "@rpath" or anything else.

It must be something in macOS Sequoia that somehow tries to load the "libpng.dylib". ImageIO kit from the OS. But why is it trying that. "libpng.*.dylib" is not even part of the system.

Maybe it is "libz.1.dylib" but have tried to compile with the one provided with macOS, but I still get this error.

**Reply by carlocab** [reply]
That seems unlikely. Is the app setting `DYLD_LIBRARY_PATH`? This will override all paths to linked libraries, even ones that aren't prefixed by `@rpath`/`@loader_path`/`@executable_path`.

**Reply by mikkeljohnsen** [reply]
It is setting DYLD_LIBRARY_PATH to where the libs that I compile is located, including the "libpng.dylib". But only to that location.

And since I need "libpng" for Gtk etc to work, then removing it is not a option.

But I do not understand why ImageIO fails on "IIOReadPlugin". Why is that loading "libpng".

**Reply by carlocab** [reply]
> It is setting DYLD_LIBRARY_PATH to where the libs that I compile is located, including the "libpng.dylib". 

If your app is built correctly, it should not need to use `DYLD_LIBRARY_PATH`. I suggest dropping that.

**Reply by mikkeljohnsen** [reply]
It has no effect to remove the DYLD_LIBRARY_PATH the app is still crashing.

**Reply by carlocab** [reply]
Then it's some other bug in your app. It shouldn't be affected by whether random libraries are installed on your system or not.

**Reply by mikkeljohnsen** [reply]
No it should'nt, but the app has been working for many many years, and only started not working on Sequoia, works fine on macOS versions before that.

So something has changed in Sequoia how, and what is loaded !!

**Reply by mikkeljohnsen** [reply]
The problem only occurs when running with "dotnet" if I run a  test program in C there is no problem.

So somehow "dotnet" is coursing the problem

**Comment by mikkeljohnsen** [comment]
So the problem is that "dotnet" links to "CoreServices" that uses "ImageIO" framework. The ImageIO framework is using a "libPng.dylib" (I assume is very old). But if another "libpng" is installed from Homebrew or any other. Then normally you would have this:

```
libpng16.16.dylib
libpng.dylib -> libpng16.16.dylib
libpng16.dylib -> libpng16.16.dylib
```

The crash is because the "libpng.dylib" symbolic link is present, then the ImageIO framework tries to load that "libpng" and it do not match the API it expected.

So the simple solution is to delete the "libpng.dylib" symbolic link.

I would think the same solution could be used here. But not sure.
